libs/base/Control/Monad/Partial.idr

```diff
@@ -0,0 +1,52 @@
+module Control.Monad.Partial
+
+%default total
+
+
+||| Delay/partiality monad. Represents partial computations which may not
+||| terminate or cover all inputs. This can be preferable to defining a partial
+||| Idris function as it allows termination to be proven as an external proof.
+public export
+data Partial : Type -> Type where
+    Now : a -> Partial a
+    Later : Inf (Partial a) -> Partial a
+
+
+public export
+Functor Partial where
+    map f (Now x) = Now (f x)
+    map f (Later x) = Later (map f x)
+
+
+public export
+Applicative Partial where
+    pure = Now
+
+    Now f <*> x = map f x
+    Later f <*> x = Later (f <*> x)
+
+
+public export
+Monad Partial where
+    Now x >>= f = f x
+    Later x >>= f = Later (x >>= f)
+
+
+||| A partial computation which never terminates (aka. undefined/bottom).
+public export
+never : Partial a
+never = Later never
+
+
+public export
+data Total : Partial a -> Type where
```

Comment:
In base such proves usually have name `Is<Property>`, like `isJust` for `Maybe` and `IsYes` for `Dec`. Maybe, this one should be names `IsTotal` to be coherent with the others

Reply:
That naming pattern seems to be primarily used for proofs that a data type is a specific constructor, whereas I would argue that this type is more similar to something like `Elem`. (Though I don't really mind what it is named)